File: ai-app-generation-frontend/src/utils/buildProjectToolResult.ts
export interface BuildProjectToolView {
  invocationStatus: 'COMPLETED' | 'BUILD_IN_PROGRESS' | 'REJECTED' | 'CANCELLED'
  success?: boolean
  attempt?: number
  maxAttempts: number
  stage?: string
  statusText: string
  errorSummary?: string
  terminal: boolean
}

type JsonRecord = Record<string, unknown>
type InvocationStatus = BuildProjectToolView['invocationStatus']

const PROTOCOL = 'vue-build-tool/v1'
const MAX_ATTEMPTS = 3
const MAX_ERROR_SUMMARY_LENGTH = 1000
const FAILURE_RESPONSE = '抱歉，系统遇到了一些问题，请您稍后重试修复'
const SUCCESS_RESPONSE = '项目已生成并构建成功。'
const INVOCATION_STATUSES = new Set<InvocationStatus>([
  'COMPLETED',
  'BUILD_IN_PROGRESS',
  'REJECTED',
  'CANCELLED',
])
const NEXT_ACTIONS = new Set(['REPAIR', 'RETRY_BUILD', 'FINAL_DIAGNOSIS', 'STOP'])
const BUILD_STAGES = new Set(['VALIDATION', 'NPM_INSTALL', 'NPM_BUILD', 'DIST_CHECK', 'SUCCESS'])
const FAILURE_KINDS = new Set(['CODE', 'DEPENDENCY', 'INFRASTRUCTURE'])

function parseRecord(raw: unknown): JsonRecord | undefined {
  if (typeof raw === 'string') {
    try {
      const parsed = JSON.parse(raw) as unknown
      return isRecord(parsed) ? parsed : undefined
    } catch {
      return undefined
    }
  }
  return isRecord(raw) ? raw : undefined
}

function isRecord(value: unknown): value is JsonRecord {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function isNullableString(value: unknown): value is string | null {
  return value === null || typeof value === 'string'
}

function isNullableBoolean(value: unknown): value is boolean | null {
  return value === null || typeof value === 'boolean'
}

function isNullableAttempt(value: unknown): value is number | null {
  return value === null || (Number.isInteger(value) && Number(value) >= 1 && Number(value) <= 3)
}

function hasCommonFields(record: JsonRecord): boolean {
  return (
    record.protocol === PROTOCOL &&
    INVOCATION_STATUSES.has(record.invocationStatus as InvocationStatus) &&
    record.maxAttempts === MAX_ATTEMPTS &&
    typeof record.message === 'string' &&
    record.message.trim().length > 0 &&
    isNullableBoolean(record.success) &&
    isNullableAttempt(record.attempt) &&
    isNullableString(record.stage) &&
    isNullableString(record.failureKind) &&
    isNullableBoolean(record.timedOut) &&
    typeof record.repairable === 'boolean' &&
    typeof record.reflectionRequired === 'boolean' &&
    isNullableString(record.nextAction) &&
    isNullableString(record.errorSummary) &&
    typeof record.terminateToolLoop === 'boolean' &&
    isNullableString(record.finalResponse)
  )
}

function isCompletedSuccess(record: JsonRecord): boolean {
  return (
    record.success === true &&
    record.stage === 'SUCCESS' &&
    record.failureKind === null &&
    record.timedOut === false &&
    record.repairable === false &&
    record.reflectionRequired === false &&
    record.nextAction === 'STOP' &&
    record.errorSummary === null &&
    record.terminateToolLoop === true &&
    record.finalResponse === SUCCESS_RESPONSE
  )
}

function expectedFailureAction(record: JsonRecord): string {
  if (record.attempt === MAX_ATTEMPTS) {
    return 'STOP'
  }
  if (record.attempt === 2) {
    return 'FINAL_DIAGNOSIS'
  }
  return record.failureKind === 'CODE' ? 'REPAIR' : 'RETRY_BUILD'
}

function isCompletedFailure(record: JsonRecord): boolean {
  if (
    record.success !== false ||
    record.stage === 'SUCCESS' ||
    typeof record.failureKind !== 'string' ||
    !FAILURE_KINDS.has(record.failureKind) ||
    typeof record.errorSummary !== 'string' ||
    !NEXT_ACTIONS.has(record.nextAction as string)
  ) {
    return false
  }
  const terminal = record.attempt === MAX_ATTEMPTS
  return (
    record.repairable === (record.attempt === 1 && record.failureKind === 'CODE') &&
    record.reflectionRequired === (Number(record.attempt) >= 2) &&
    record.nextAction === expectedFailureAction(record) &&
    record.terminateToolLoop === terminal &&
    (terminal ? record.finalResponse === FAILURE_RESPONSE : record.finalResponse === null)
  )
}

function completedStatusText(record: JsonRecord): string {
  if (record.success === true) {
    return `第 ${record.attempt} 次构建成功`
  }
  if (record.nextAction === 'REPAIR') {
    return '第 1 次构建失败，正在进行最小修复'
  }
  if (record.nextAction === 'RETRY_BUILD') {
    return '第 1 次构建失败，正在直接重试构建'
  }
  if (record.nextAction === 'FINAL_DIAGNOSIS') {
    return '第 2 次构建失败，正在进行最终诊断'
  }
  return `第 ${record.attempt} 次构建失败，已停止自动修复`
}

function parseCompleted(record: JsonRecord): BuildProjectToolView | undefined {
  if (
    typeof record.success !== 'boolean' ||
    typeof record.attempt !== 'number' ||
    typeof record.stage !== 'string' ||
    !BUILD_STAGES.has(record.stage) ||
    typeof record.timedOut !== 'boolean' ||
    typeof record.nextAction !== 'string' ||
    (!isCompletedSuccess(record) && !isCompletedFailure(record))
  ) {
    return undefined
  }
  return {
    invocationStatus: 'COMPLETED',
    success: record.success,
    attempt: record.attempt,
    maxAttempts: MAX_ATTEMPTS,
    stage: record.stage,
    statusText: completedStatusText(record),
    errorSummary:
      typeof record.errorSummary === 'string'
        ? record.errorSummary.slice(0, MAX_ERROR_SUMMARY_LENGTH)
        : undefined,
    terminal: record.terminateToolLoop as boolean,
  }
}

function parseCancelled(record: JsonRecord): BuildProjectToolView | undefined {
  const hasAttemptAndStage = typeof record.attempt === 'number' && typeof record.stage === 'string'
  const hasNeither = record.attempt === null && record.stage === null
  if (
    record.success !== null ||
    record.failureKind !== null ||
    record.timedOut !== null ||
    record.repairable !== false ||
    record.reflectionRequired !== false ||
    record.nextAction !== 'STOP' ||
    record.errorSummary !== null ||
    record.terminateToolLoop !== true ||
    record.finalResponse !== FAILURE_RESPONSE ||
    (hasAttemptAndStage && !BUILD_STAGES.has(record.stage as string)) ||
    (!hasAttemptAndStage && !hasNeither)
  ) {
    return undefined
  }
  return {
    invocationStatus: 'CANCELLED',
    success: undefined,
    attempt: hasAttemptAndStage ? (record.attempt as number) : undefined,
    maxAttempts: MAX_ATTEMPTS,
    stage: hasAttemptAndStage ? (record.stage as string) : undefined,
    statusText: record.message as string,
    terminal: true,
  }
}

function parseTransient(record: JsonRecord): BuildProjectToolView | undefined {
  if (
    record.success !== null ||
    record.attempt !== null ||
    record.stage !== null ||
    record.failureKind !== null ||
    record.timedOut !== null ||
    record.repairable !== false ||
    record.reflectionRequired !== false ||
    record.nextAction !== null ||
    record.errorSummary !== null
  ) {
    return undefined
  }
  const status = record.invocationStatus as 'BUILD_IN_PROGRESS' | 'REJECTED'
  if (status === 'BUILD_IN_PROGRESS' && (record.terminateToolLoop || record.finalResponse !== null)) {
    return undefined
  }
  if (
    status === 'REJECTED' &&
    (record.terminateToolLoop !== (record.finalResponse !== null) ||
      (record.terminateToolLoop && record.finalResponse !== FAILURE_RESPONSE))
  ) {
    return undefined
  }
  return {
    invocationStatus: status,
    success: undefined,
    maxAttempts: MAX_ATTEMPTS,
    statusText: record.message as string,
    terminal: record.terminateToolLoop as boolean,
  }
}

export function parseBuildProjectToolResult(raw: unknown): BuildProjectToolView | undefined {
  const record = parseRecord(raw)
  if (!record || !hasCommonFields(record)) {
    return undefined
  }
  if (record.invocationStatus === 'COMPLETED') {
    return parseCompleted(record)
  }
  if (record.invocationStatus === 'CANCELLED') {
    return parseCancelled(record)
  }
  return parseTransient(record)
}

File: src/main/java/com/lyw/appgeneration/ai/tools/BuildProjectToolResult.java
package com.lyw.appgeneration.ai.tools;

import com.lyw.appgeneration.core.builder.BuildStage;
import com.lyw.appgeneration.core.builder.VueBuildFailureKind;

import java.util.Objects;

/** Vue 构建工具返回给模型的受信结构化协议。 */
public record BuildProjectToolResult(
        String protocol,
        BuildInvocationStatus invocationStatus,
        Boolean success,
        Integer attempt,
        int maxAttempts,
        BuildStage stage,
        VueBuildFailureKind failureKind,
        Boolean timedOut,
        boolean repairable,
        boolean reflectionRequired,
        BuildNextAction nextAction,
        String message,
        String errorSummary,
        boolean terminateToolLoop,
        String finalResponse
) {

    public static final String PROTOCOL = "vue-build-tool/v1";
    public static final int MAX_ATTEMPTS = 3;
    public static final String SUCCESS_RESPONSE = "项目已生成并构建成功。";
    public static final String FAILURE_RESPONSE = "抱歉，系统遇到了一些问题，请您稍后重试修复";

    public BuildProjectToolResult {
        if (!PROTOCOL.equals(protocol) || maxAttempts != MAX_ATTEMPTS) {
            throw new IllegalArgumentException("Vue 构建工具协议版本不受支持");
        }
        Objects.requireNonNull(invocationStatus, "invocationStatus 不能为空");
        Objects.requireNonNull(message, "message 不能为空");
        validateState(invocationStatus, success, attempt, stage, failureKind, timedOut,
                repairable, reflectionRequired, nextAction, errorSummary,
                terminateToolLoop, finalResponse);
    }

    public static BuildProjectToolResult completedSuccess(int attempt) {
        return completed(true, attempt, BuildStage.SUCCESS, null, false,
                false, false, BuildNextAction.STOP, "构建成功", null,
                true, SUCCESS_RESPONSE);
    }

    public static BuildProjectToolResult completedFailure(
            int attempt,
            BuildStage stage,
            VueBuildFailureKind failureKind,
            boolean timedOut,
            String errorSummary) {
        BuildNextAction action = failureAction(attempt, failureKind);
        boolean repairable = attempt < MAX_ATTEMPTS
                && failureKind == VueBuildFailureKind.CODE;
        boolean reflectionRequired = attempt == 2;
        boolean terminate = attempt >= MAX_ATTEMPTS;
        return completed(false, attempt, stage, failureKind, timedOut,
                repairable, reflectionRequired, action, failureMessage(attempt, action),
                errorSummary, terminate, terminate ? FAILURE_RESPONSE : null);
    }

    public static BuildProjectToolResult buildInProgress() {
        return transientResult(BuildInvocationStatus.BUILD_IN_PROGRESS,
                "当前已有构建正在执行", false, null);
    }

    public static BuildProjectToolResult mutationRequired(String message) {
        return transientResult(
                BuildInvocationStatus.REJECTED, message, false, BuildNextAction.REPAIR);
    }

    public static BuildProjectToolResult rejected(String message) {
        return transientResult(
                BuildInvocationStatus.REJECTED, message, true, BuildNextAction.STOP);
    }

    public static BuildProjectToolResult cancelled(
            Integer attempt, BuildStage stage, String message) {
        return new BuildProjectToolResult(PROTOCOL, BuildInvocationStatus.CANCELLED,
                null, attempt, MAX_ATTEMPTS, stage, null, null,
                false, false, BuildNextAction.STOP, message, null,
                true, null);
    }

    private static BuildProjectToolResult completed(
            boolean success,
            int attempt,
            BuildStage stage,
            VueBuildFailureKind failureKind,
            boolean timedOut,
            boolean repairable,
            boolean reflectionRequired,
            BuildNextAction nextAction,
            String message,
            String errorSummary,
            boolean terminate,
            String finalResponse) {
        return new BuildProjectToolResult(PROTOCOL, BuildInvocationStatus.COMPLETED,
                success, attempt, MAX_ATTEMPTS, stage, failureKind, timedOut,
                repairable, reflectionRequired, nextAction, message, errorSummary,
                terminate, finalResponse);
    }

    private static BuildProjectToolResult transientResult(
            BuildInvocationStatus status,
            String message,
            boolean terminate,
            BuildNextAction nextAction) {
        return new BuildProjectToolResult(PROTOCOL, status, null, null,
                MAX_ATTEMPTS, null, null, null, false, false, nextAction,
                message, null, terminate, null);
    }

    private static BuildNextAction failureAction(
            int attempt, VueBuildFailureKind failureKind) {
        if (attempt >= MAX_ATTEMPTS) {
            return BuildNextAction.STOP;
        }
        if (attempt == 2) {
            return BuildNextAction.FINAL_DIAGNOSIS;
        }
        return failureKind == VueBuildFailureKind.CODE
                ? BuildNextAction.REPAIR : BuildNextAction.RETRY_BUILD;
    }

    private static String failureMessage(int attempt, BuildNextAction action) {
        return switch (action) {
            case REPAIR -> "第 1 次构建失败，请进行最小代码修复";
            case RETRY_BUILD -> "第 1 次构建失败，请直接重试构建";
            case FINAL_DIAGNOSIS -> "第 2 次构建失败，请进行最终诊断";
            case STOP -> "第 " + attempt + " 次构建失败，已停止自动修复";
        };
    }

    private static void validateState(
            BuildInvocationStatus status,
            Boolean success,
            Integer attempt,
            BuildStage stage,
            VueBuildFailureKind failureKind,
            Boolean timedOut,
            boolean repairable,
            boolean reflectionRequired,
            BuildNextAction nextAction,
            String errorSummary,
            boolean terminateToolLoop,
            String finalResponse) {
        if (status == BuildInvocationStatus.COMPLETED) {
            if (success == null || attempt == null || stage == null
                    || timedOut == null || nextAction == null) {
                throw new IllegalArgumentException("已完成构建缺少结果字段");
            }
            if (attempt < 1 || attempt > MAX_ATTEMPTS) {
                throw new IllegalArgumentException("构建次数超出协议范围");
            }
            validateCompletedState(success, attempt, stage, failureKind, timedOut,
                    repairable, reflectionRequired, nextAction, errorSummary,
                    terminateToolLoop, finalResponse);
            return;
        }
        if (status == BuildInvocationStatus.CANCELLED) {
            if (success != null || timedOut != null || nextAction != BuildNextAction.STOP
                    || failureKind != null || repairable || reflectionRequired
                    || errorSummary != null || !terminateToolLoop
                    || (attempt == null) != (stage == null)
                    || finalResponse != null) {
                throw new IllegalArgumentException("取消结果字段组合不合法");
            }
            if (attempt != null && (attempt < 1 || attempt > MAX_ATTEMPTS)) {
                throw new IllegalArgumentException("取消结果的构建次数超出协议范围");
            }
            return;
        }
        if (success != null || attempt != null || stage != null
                || failureKind != null || timedOut != null || repairable
                || reflectionRequired || errorSummary != null) {
            throw new IllegalArgumentException("未完成调用不能伪造构建结果字段");
        }
        if (status == BuildInvocationStatus.BUILD_IN_PROGRESS
                && (nextAction != null || terminateToolLoop || finalResponse != null)) {
            throw new IllegalArgumentException("构建占用状态不能终止工具循环");
        }
        if (status == BuildInvocationStatus.REJECTED
                && (finalResponse != null
                || terminateToolLoop != (nextAction == BuildNextAction.STOP)
                || (!terminateToolLoop && nextAction != BuildNextAction.REPAIR))) {
            throw new IllegalArgumentException("拒绝结果的终止字段组合不合法");
        }
    }

    private static void validateCompletedState(
            boolean success,
            int attempt,
            BuildStage stage,
            VueBuildFailureKind failureKind,
            boolean timedOut,
            boolean repairable,
            boolean reflectionRequired,
            BuildNextAction nextAction,
            String errorSummary,
            boolean terminateToolLoop,
            String finalResponse) {
        if (success) {
            if (stage != BuildStage.SUCCESS || failureKind != null || timedOut
                    || repairable || reflectionRequired || nextAction != BuildNextAction.STOP
                    || errorSummary != null || !terminateToolLoop
                    || !SUCCESS_RESPONSE.equals(finalResponse)) {
                throw new IllegalArgumentException("成功构建结果字段组合不合法");
            }
            return;
        }
        BuildNextAction expectedAction = failureAction(attempt, failureKind);
        boolean expectedRepairable = attempt < MAX_ATTEMPTS
                && failureKind == VueBuildFailureKind.CODE;
        boolean expectedReflection = attempt == 2;
        boolean expectedTerminate = attempt == MAX_ATTEMPTS;
        if (stage == BuildStage.SUCCESS || failureKind == null || errorSummary == null
                || repairable != expectedRepairable
                || reflectionRequired != expectedReflection
                || nextAction != expectedAction
                || terminateToolLoop != expectedTerminate
                || (expectedTerminate
                ? !FAILURE_RESPONSE.equals(finalResponse)
                : finalResponse != null)) {
            throw new IllegalArgumentException("失败构建结果字段组合不合法");
        }
    }

    public enum BuildInvocationStatus {
        COMPLETED,
        BUILD_IN_PROGRESS,
        REJECTED,
        CANCELLED
    }

    public enum BuildNextAction {
        REPAIR,
        RETRY_BUILD,
        FINAL_DIAGNOSIS,
        STOP
    }
}
